packages/devtools/lib/src/timeline/frame_flame_chart.dart

```diff
@@ -407,3 +430,142 @@ class FlameChartItem {
         colorToCss(selected ? selectedTextColor : defaultTextColor);
   }
 }
+
+class TimelineGrid extends CoreElement {
+  TimelineGrid(this._frameDurationMs, this._flameChartWidth)
+      : super('div', classes: 'flame-chart-grid') {
+    flex();
+    layoutHorizontal();
+    _drawGrid(baseGridInterval);
+  }
+
+  static const baseGridInterval = 150;
+
+  final num _frameDurationMs;
+
+  num _zoomLevel = 1;
+
+  num _flameChartWidth;
+
+  num get _flameChartWidthWithInsets => _flameChartWidth + 2 * flameChartInset;
+
+  final List<TimelineGridItem> _gridItems = [];
+
+  void _drawGrid(num interval) {
+    // Draw the first grid item since it will have a different width than the
+    // rest.
+    final gridItem = TimelineGridItem(0, flameChartInset, 0);
+    _gridItems.add(gridItem);
+    add(gridItem);
+
+    num left = flameChartInset;
+
+    while (left + interval < _flameChartWidthWithInsets) {
+      final timestamp = getTimestampForPosition(left + interval);
+      final gridItem = TimelineGridItem(left, interval, timestamp);
+
+      _gridItems.add(gridItem);
+      add(gridItem);
+
+      left += interval;
+    }
+  }
+
+  void _clear() {
+    clear();
+    _gridItems.clear();
+  }
+
+  num getTimestampForPosition(num gridItemEnd) {
+    return (gridItemEnd - flameChartInset) /
+        _flameChartWidth *
+        _frameDurationMs;
+  }
+
+  void updateForZoom(num newZoomLevel, num newFlameChartWidth) {
+    if (_zoomLevel == newZoomLevel) {
+      return;
+    }
+
+    _flameChartWidth = newFlameChartWidth;
+    element.style.width = '${_flameChartWidthWithInsets}px';
+
+    final log2ZoomLevel = log2(_zoomLevel);
+    final log2NewZoomLevel = log2(newZoomLevel);
+
+    final gridZoomFactor = math.pow(2, log2NewZoomLevel);
+    final gridIntervalPx = baseGridInterval / gridZoomFactor;
+
+    /// The physical pixel width of the grid interval at [newZoomLevel].
+    final zoomedGridIntervalPx = gridIntervalPx * newZoomLevel;
+
```

Comment:
@jacob314 any opinion on the naming of these vars?

Reply:
the names are good enough for now.